gcloud-java-dns/src/main/java/com/google/gcloud/spi/DefaultDnsRpc.java

```diff
@@ -0,0 +1,178 @@
+    this.dns = new Dns.Builder(transport, new JacksonFactory(), initializer)
+        .setRootUrl(options.host())
+        .setApplicationName(options.applicationName())
+        .build();
+    this.options = options;
+  }
+
+  @Override
+  public ManagedZone create(ManagedZone zone) throws DnsException {
+    try {
+      return dns.managedZones().create(this.options.projectId(), zone).execute();
+    } catch (IOException ex) {
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public ManagedZone getZone(String zoneName, Map<Option, ?> options) throws DnsException {
+    // just fields option
+    try {
+      return dns.managedZones().get(this.options.projectId(), zoneName)
+          .setFields(FIELDS.getString(options)).execute();
+    } catch (IOException ex) {
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public Tuple<String, Iterable<ManagedZone>> listZones(Map<Option, ?> options)
+      throws DnsException {
+    // fields, page token, page size
+    try {
+      ManagedZonesListResponse zoneList = dns.managedZones().list(this.options.projectId())
+          .setFields(FIELDS.getString(options))
+          .setMaxResults(PAGE_SIZE.getInt(options))
+          .setPageToken(PAGE_TOKEN.getString(options))
+          .execute();
+      return Tuple.<String, Iterable<ManagedZone>>of(zoneList.getNextPageToken(),
+          zoneList.getManagedZones());
+    } catch (IOException ex) {
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public boolean deleteZone(String zoneName) throws DnsException {
+    try {
+      dns.managedZones().delete(this.options.projectId(), zoneName).execute();
+      return true;
+    } catch (IOException ex) {
+      DnsException serviceException = translate(ex);
+      if (serviceException.code() == HTTP_NOT_FOUND) {
+        return false;
+      }
+      throw serviceException;
+    }
+  }
+
+  @Override
+  public Tuple<String, Iterable<ResourceRecordSet>> listDnsRecords(String zoneName,
+      Map<Option, ?> options) throws DnsException {
+    // options are fields, page token, dns name, type
+    try {
+      ResourceRecordSetsListResponse response = dns.resourceRecordSets()
+          .list(this.options.projectId(), zoneName)
+          .setFields(FIELDS.getString(options))
+          .setPageToken(PAGE_TOKEN.getString(options))
+          .setMaxResults(PAGE_SIZE.getInt(options))
+          .setName(DNS_NAME.getString(options))
+          .setType(DNS_TYPE.getString(options))
+          .execute();
+      return Tuple.<String, Iterable<ResourceRecordSet>>of(response.getNextPageToken(),
+          response.getRrsets());
+    } catch (IOException ex) {
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public Project getProject(Map<Option, ?> options) throws DnsException {
+    try {
+      return dns.projects().get(this.options.projectId())
+          .setFields(FIELDS.getString(options)).execute();
+    } catch (IOException ex) {
+      throw translate(ex);
```

Comment:
(I originally typed this below) TL;DR: I would do an exception.

I think this should never happen in the first place. However, this is kind of the scenario when the parent container does not exist, is it not? Furthermore, it feels consistent with other parts of the library. Say that you want to fetch a bucket from non-existent project using Storage, you will get an exception as well, right?

The service call itself gives us 400 error.

For now, I at least specified it in the doc.